### PR TITLE
Tests and other tweaks to shared/settings.js

### PR DESCRIPTION
**Pull request by seanh**
Another little thing that I've extracted out of config code refactoring work:

* Rewrite the `shared/settings.js` unit tests with more detailed ones

* Have `shared/settings.js` export an object with a `jsonConfigsFrom(document)` function rather than just exporting a top-level function.

   The way I'm intending for `shared/settings.js` to be used is as a container for helper / utility functions that are called by both the `annotator` and the `sidebar` config code, and I expect this to contain more than just one function in the future, so I'm making space for it.

   Also "settings" in the client can be read from lots of places: the URL, js-hypothesis-config tags in the host page, or in the app.html, window.hypothesisConfig() in the host page, the annotator+html link... This `jsonConfigsFrom(document)` is really just a utility function for parsing JSON tags out of (any) document, it doesn't return "the settings" as a whole, so I think it needs a more specific name than `settings()`

* Also rewrote the function's docstring

I'd like to do more work to fix a couple of issues that the tests found:

* Stop this function from throwing JSON parsing errors. In every case where we use the function I think the desired behaviour would be to catch and log the error (in practice in some cases we do that and in some cases we don't catch it and crash), so the function should do the catch and log itself

* Stop this function from returning nonsense when a json tag doesn't contain a top-level object, as a few of the tests show

But I've run out of time for this week so I'll submit a separate pr for those changes later

### CODE REVIEW

File: src/annotator/config.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ function config(window_) {
 
   // Parse config from `<script class="js-hypothesis-config">` tags
   try {
-    Object.assign(options, settings(window_.document));
+    Object.assign(options, settings.jsonConfigsFrom(window_.document));
```
Comment (seanh) [inline src/annotator/config.js]: I think this is _much_ clearer, `jsonConfigsFrom(document)` tells me that this is reading the JSON objs from the document, without me having to go look in `settings.js`

File: src/shared/settings.js [hunk 2]
```diff
@@ -12,17 +12,22 @@ function assign(dest, src) {
 }
 
 /**
- * Return application configuration information from the host page.
```
Comment (seanh) [inline src/shared/settings.js]: This was actually wrong - the function isn't always used to read settings from the host page

File: src/shared/settings.js [hunk 3]
```diff
@@ -12,17 +12,22 @@ function assign(dest, src) {
...
+ * Return a parsed `js-hypothesis-config` object from the document, or `{}`.
  *
- * Exposes shared application settings, read from script tags with the
```
Comment (seanh) [inline src/shared/settings.js]: This is also misleading. It's a shared utility function, yes. But "shared application settings" is misleading. Not every setting read from a js-hypothesis-config is actually shared between the three legs of the app (annotator, boot and sidebar), nor is every setting that _is_ shared read from this function.

File: src/shared/settings.js [hunk 4]
```diff
@@ -12,17 +12,22 @@ function assign(dest, src) {
...
- * If there are multiple such tags, the configuration from each is merged.
+ * If there are no `js-hypothesis-config` tags in the document then return
+ * `{}`.
  *
- * @param {Document|Element} document - The root element to search for
- *                                      <script> settings tags.
+ * If there are multiple `js-hypothesis-config` tags in the document then merge
+ * them into a single returned object (when multiple scripts contain the same
+ * setting names, scripts further down in the document override those further
+ * up).
```
Comment (seanh) [inline src/shared/settings.js]: I've left the exception that this function throws (on invalid JSON) still undocumented, as I intend to send a PR to stop iut from throwing the exception anyway

File: src/shared/test/settings-test.js [hunk 5]
```diff
@@ -2,33 +2,128 @@
...
+        assert.deepEqual(jsonConfigsFrom(document), {});
+      });
+    });
+
+    context("when there's a JSON script with a top-level array", function() {
+      beforeEach('add a JSON script containing a top-level array', function() {
+        appendJSHypothesisConfig(document, '["a", "b", "c"]');
+      });
+
+      it('returns the array, parsed into an object', function() {
```
Comment (seanh) [inline src/shared/test/settings-test.js]: We should fix this in a separate pr, it should ignore top-level arrays (but log a warning)

File: src/shared/test/settings-test.js [hunk 6]
```diff
@@ -2,33 +2,128 @@
...
+        );
+      });
+    });
+
+    context("when there's a JSON script with a top-level string", function() {
+      beforeEach('add a JSON script with a top-level string', function() {
+        appendJSHypothesisConfig(document, '"hi"');
+      });
+
+      it('returns the string, parsed into an object', function() {
```
Comment (seanh) [inline src/shared/test/settings-test.js]: We should fix this in a separate pr, it should ignore top-level strings (but log a warning)

File: src/shared/test/settings-test.js [hunk 7]
```diff
@@ -2,33 +2,128 @@
...
+        assert.deepEqual(jsonConfigsFrom(document), {0: 'h', 1: 'i'});
+      });
+    });
+
+    context("when there's a JSON script containing invalid JSON", function() {
+      beforeEach('add a JSON script containing invalid JSON', function() {
+        appendJSHypothesisConfig(document, 'this is not valid json');
+      });
+
+      it('throws a SyntaxError', function() {
```
Comment (seanh) [inline src/shared/test/settings-test.js]: We should fix this in a follow up pr: it should log a warning, and ignore the invalid json, not throw

**Comment by robertknight** [comment]
> Stop this function from throwing JSON parsing errors. In every case where we use the function I think the desired behaviour would be to catch and log the error

Unhandled errors in the sidebar code get reported to us via Raven, so if you _catch_ the error in the context of the sidebar and log an error, that could actually have the effect of _hiding_ the problem from us.

File: src/shared/settings.js [hunk 8]
```diff
@@ -12,17 +12,22 @@ function assign(dest, src) {
...
+ *
+ * @param {Document|Element} document - The root element to search for.
```
Comment (robertknight) [inline src/shared/settings.js]: The end of the sentence ("... search for script tags") is missing here.

File: src/shared/test/settings-test.js [hunk 9]
```diff
@@ -2,33 +2,128 @@
...
+      el.textContent = jsonString;
+      el.classList.add('js-hypothesis-config');
+      el.classList.add('js-settings-test');
+      document_.body.appendChild(el);
+    }
+
+    afterEach('remove js-hypothesis-config tags', function() {
+      var elements = document.querySelectorAll('.js-settings-test');
+      for (var i=0; i < elements.length; i++) {
+        elements[i].parentNode.removeChild(elements[i]);
```
Comment (robertknight) [inline src/shared/test/settings-test.js]: You can shorten this to `elements[i].remove()`. Some older code uses `removeChild()` because `Element#remove` was not supported in older versions of PhantomJS.

File: src/shared/test/settings-test.js (hunk 6 above)
Comment (robertknight) [inline src/shared/test/settings-test.js]: Heh. LOLJS :)

**Comment by seanh** [comment]
Done both suggestions
